// @umbraco/headless-client/Responses/Form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FormFieldIndicationType = exports.FormConditionActionType = exports.FormConditionLogicType = exports.FormConditionRuleOperatorType = void 0;
/**
 * @public
 */
var FormConditionRuleOperatorType;
(function (FormConditionRuleOperatorType) {
    FormConditionRuleOperatorType["Is"] = "IS";
    FormConditionRuleOperatorType["IsNot"] = "IS_NOT";
    FormConditionRuleOperatorType["GreaterThen"] = "GREATER_THEN";
    FormConditionRuleOperatorType["LessThen"] = "LESS_THEN";
    FormConditionRuleOperatorType["Contains"] = "CONTAINS";
    FormConditionRuleOperatorType["StartsWith"] = "STARTS_WITH";
    FormConditionRuleOperatorType["EndsWith"] = "ENDS_WITH";
})(FormConditionRuleOperatorType = exports.FormConditionRuleOperatorType || (exports.FormConditionRuleOperatorType = {}));
/**
 * @public
 */
var FormConditionLogicType;
(function (FormConditionLogicType) {
    FormConditionLogicType["All"] = "ALL";
    FormConditionLogicType["Any"] = "ANY";
})(FormConditionLogicType = exports.FormConditionLogicType || (exports.FormConditionLogicType = {}));
/**
 * @public
 */
var FormConditionActionType;
(function (FormConditionActionType) {
    FormConditionActionType["Show"] = "SHOW";
    FormConditionActionType["Hide"] = "HIDE";
})(FormConditionActionType = exports.FormConditionActionType || (exports.FormConditionActionType = {}));
/**
 * @public
 */
var FormFieldIndicationType;
(function (FormFieldIndicationType) {
    FormFieldIndicationType["NoIndicator"] = "NO_INDICATOR";
    FormFieldIndicationType["MarkMandatoryFields"] = "MARK_MANDATORY_FIELDS";
    FormFieldIndicationType["MarkOptionalFields"] = "MARK_OPTIONAL_FIELDS";
})(FormFieldIndicationType = exports.FormFieldIndicationType || (exports.FormFieldIndicationType = {}));
