// @umbraco/headless-client/Clients/Management/ManagementClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagementClient = void 0;
var ContentManagementClient_1 = require("./ContentManagementClient");
var MediaManagementClient_1 = require("./MediaManagementClient");
var MemberManagementClient_1 = require("./MemberManagementClient");
var Endpoints_1 = require("../../Endpoints");
/**
 * ManagementClient is used to access the Content Management API.
 * @public
 *
 * @example
 * The {@link ManagementClient} must be accessed through {@link Client}.
 *
 * ```typescript
 * import { Client } from '@umbraco/headless-client'
 *
 * const client = new Client({
 *  projectAlias: '<your-project-alias>',
 *  apiKey: '<your-api-key>',
 *  language: '<iso-code>',
 * })
 *
 * const managementClient = client.management
 * ```
 */
var ManagementClient = /** @class */ (function () {
    /**
     * @internal
     */
    function ManagementClient(client) {
        var _this = this;
        this.client = client;
        this.makeRequest = function (endpoint, data) { return __awaiter(_this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.client.makeRequest(endpoint, data)];
            });
        }); };
        /**
         * The Content client for the Content Management API.
         * See {@link ContentManagementClient}
         */
        this.content = new ContentManagementClient_1.ContentManagementClient(this.client);
        /**
         * The Media client for the Media Management API.
         * See {@link MediaManagementClient}
         */
        this.media = new MediaManagementClient_1.MediaManagementClient(this.client);
        /**
         * The Member client for the Member Management API.
         * See {@link MemberManagementClient}
         */
        this.member = new MemberManagementClient_1.MemberManagementClient(this.client);
    }
    Object.defineProperty(ManagementClient.prototype, "contentType", {
        /**
         * ContentType API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch all content types
                 */
                all: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.contentType.all())];
                }); }); },
                /**
                 * Find content type by alias
                 * @param alias Alias for the content type
                 */
                byAlias: function (alias) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.contentType.byAlias(alias))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "mediaType", {
        /**
         * Media API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch all media types
                 */
                all: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.mediaType.all())];
                }); }); },
                /**
                 * Find media type by alias
                 * @param alias Alias of the media type querying for
                 */
                byAlias: function (alias) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.mediaType.byAlias(alias))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "language", {
        /**
         * Language API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch all languages
                 */
                all: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.language.all())];
                }); }); },
                /**
                 * Find language by ISO code
                 * @param id ISO Code for the language (e.g. en-US)
                 */
                byISOCode: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.language.byISOCode(id))];
                }); }); },
                /**
                 * Create a language
                 * @param data Data for creating language object
                 */
                create: function (data) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.language.create(), data)];
                }); }); },
                /**
                 * Update a language
                 * @param id ISO Code for the language (e.g. en-US)
                 * @param data Data for updating language object
                 */
                update: function (id, data) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.language.update(id), data)];
                }); }); },
                /**
                 * Delete a language
                 * @param id ISO Code for the language (e.g. en-US)
                 */
                delete: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.language.delete(id))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "relation", {
        /**
         * Relation API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Find relation by id
                 * @param id GUID part of an Umbraco UDI
                 */
                byId: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relation.byId(id))];
                }); }); },
                /**
                 * Find relation by alias
                 * @param alias Alias of the relation querying for
                 */
                byAlias: function (alias) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relation.byAlias(alias))];
                }); }); },
                /**
                 * Fetch child for relation with id
                 * @param id GUID part of an Umbraco UDI
                 */
                byChild: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relation.byChild(id))];
                }); }); },
                /**
                 * Fetch parent for relation with id
                 * @param id GUID part of an Umbraco UDI
                 */
                byParent: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relation.byParent(id))];
                }); }); },
                /**
                 * Create a relation
                 * @param data Data for creating relation object
                 */
                create: function (data) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relation.create(), data)];
                }); }); },
                /**
                 * Delete relation with id
                 * @param id GUID part of an Umbraco UDI
                 */
                delete: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relation.delete(id))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "relationType", {
        /**
         * RelationType API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch relation type by alias
                 * @param alias Alias for the relation type queryed for
                 */
                byAlias: function (alias) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.relationType.byAlias(alias))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "memberGroup", {
        /**
         * MemberGroup API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch member group by name
                 * @param name The name of the group
                 */
                byName: function (name) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.memberGroup.byName(name))];
                }); }); },
                /**
                 * Create a member group
                 * @param data Data for creating a member group
                 */
                create: function (data) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.memberGroup.create(), data)];
                }); }); },
                /**
                 * Delete member group
                 * @param name Name of the member group to be removed
                 */
                delete: function (name) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.memberGroup.delete(name))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "memberType", {
        /**
         * MemberType API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch all member types
                 */
                all: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.memberType.all())];
                }); }); },
                /**
                 * Find by alias
                 * @param alias Alias for the member type to be found.
                 */
                byAlias: function (alias) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.memberType.byAlias(alias))];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ManagementClient.prototype, "forms", {
        /**
         * Forms API
         */
        get: function () {
            var _this = this;
            return {
                /**
                 * Fetch all forms
                 */
                all: function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.forms.all())];
                }); }); },
                /**
                 * Get form by id
                 */
                byId: function (id) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.forms.byId(id))];
                }); }); },
                /**
                 * Submit a new form entry
                 */
                submitEntry: function (formId, data) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                    return [2 /*return*/, this.makeRequest(Endpoints_1.Endpoints.management.forms.submitEntry(formId), data)];
                }); }); }
            };
        },
        enumerable: false,
        configurable: true
    });
    return ManagementClient;
}());
exports.ManagementClient = ManagementClient;
