// @umbraco/headless-client/ApiRequest.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApiRequest = void 0;
var Endpoint_1 = require("./Endpoint");
var APIRequestError_1 = require("./APIRequestError");
var axios_1 = __importDefault(require("axios"));
var form_data_1 = __importDefault(require("form-data"));
/** @internal */
var ApiRequest = /** @class */ (function () {
    function ApiRequest(options, endpoint, data) {
        var _this = this;
        this.options = options;
        this.endpoint = endpoint;
        this.data = data;
        this.promise = function () { return __awaiter(_this, void 0, void 0, function () {
            var headers, path, url, requestInit, method, token, response, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        headers = {
                            'Content-Type': 'application/json',
                            Accept: 'application/json+hal',
                            'api-version': '2.2'
                        };
                        if ('projectAlias' in this.options) {
                            headers['umb-project-alias'] = this.options.projectAlias;
                        }
                        if (this.endpoint.source === Endpoint_1.EndpointSource.CDN && this.options.language) {
                            headers['Accept-Language'] = this.options.language;
                        }
                        if ('apiKey' in this.options && this.options.apiKey) {
                            headers['api-key'] = this.options.apiKey;
                        }
                        path = this.endpoint.getPath();
                        url = 'https://cdn.umbraco.io';
                        if (this.endpoint.source === Endpoint_1.EndpointSource.ContentManagement) {
                            url = 'apiProxyUrl' in this.options
                                ? this.options.apiProxyUrl
                                : 'https://api.umbraco.io';
                        }
                        else if ('cdnProxyUrl' in this.options) {
                            url = this.options.cdnProxyUrl;
                        }
                        else if (this.options.preview) {
                            url = 'https://preview.umbraco.io';
                        }
                        url = url.endsWith('/') ? "" + url + path.substr(1) : "" + url + path;
                        requestInit = {
                            url: url,
                            method: this.endpoint.method,
                            headers: {}
                        };
                        method = this.endpoint.method.toLowerCase();
                        if ((method === 'post' || method === 'put') && !!this.data) {
                            if (this.data instanceof form_data_1.default) {
                                headers['Content-Type'] = "multipart/form-data; boundary=" + this.data.getBoundary();
                                requestInit.data = this.data;
                            }
                            else if (this.data instanceof URLSearchParams) {
                                headers['Content-Type'] = 'application/x-www-form-urlencoded';
                                requestInit.data = this.data;
                            }
                            else {
                                requestInit.data = JSON.stringify(this.data);
                            }
                        }
                        requestInit.headers = headers;
                        if ('accessTokenResolver' in this.options) {
                            token = this.options.accessTokenResolver(requestInit);
                            if (token) {
                                requestInit.headers.Authorization = "Bearer " + token;
                            }
                        }
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, axios_1.default(requestInit)];
                    case 2:
                        response = _a.sent();
                        return [2 /*return*/, response.data];
                    case 3:
                        err_1 = _a.sent();
                        throw new APIRequestError_1.APIRequestError(err_1.message, err_1.response);
                    case 4: return [2 /*return*/];
                }
            });
        }); };
    }
    return ApiRequest;
}());
exports.ApiRequest = ApiRequest;
