// @umbraco/headless-client/Endpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Endpoint = exports.EndpointSource = void 0;
/** @internal */
var EndpointSource;
(function (EndpointSource) {
    EndpointSource[EndpointSource["CDN"] = 0] = "CDN";
    EndpointSource[EndpointSource["Media"] = 1] = "Media";
    EndpointSource[EndpointSource["ContentManagement"] = 2] = "ContentManagement";
})(EndpointSource = exports.EndpointSource || (exports.EndpointSource = {}));
/**
 * This class describes how and endpoint might will look,
 * it's not possible to change value
 * @internal
 */
var Endpoint = /** @class */ (function () {
    function Endpoint(source, path, urlParams, method, options) {
        var _this = this;
        this.source = source;
        this.path = path;
        this.urlParams = urlParams;
        this.method = method;
        this.options = options;
        /**
         * Replace path with urlParams
         */
        this.getPath = function () {
            var keys = Object.keys(_this.urlParams);
            if (keys.length === 0) {
                return _this.path;
            }
            var path = _this.path;
            keys.forEach(function (key) {
                var value = _this.urlParams[key];
                var regEx = new RegExp("{" + key + "}");
                path = path.replace(regEx, value);
            });
            var params = new URLSearchParams();
            if (_this.options) {
                if ('pageSize' in _this.options && typeof _this.options.pageSize === 'number') {
                    params.append('pageSize', _this.options.pageSize.toString());
                }
                if ('page' in _this.options && typeof _this.options.page === 'number') {
                    params.append('page', _this.options.page.toString());
                }
                if ('depth' in _this.options && typeof _this.options.depth === 'number') {
                    params.append('depth', _this.options.depth.toString());
                }
                if ('hyperlinks' in _this.options && typeof _this.options.hyperlinks === 'boolean') {
                    params.append('hyperlinks', _this.options.hyperlinks ? 'true' : 'false');
                }
                if ('contentType' in _this.options && typeof _this.options.contentType === 'string') {
                    params.append('contentType', _this.options.contentType);
                }
                if ('culture' in _this.options && typeof _this.options.culture === 'string') {
                    params.append('culture', _this.options.culture);
                }
            }
            var queryString = params.toString();
            if (queryString) {
                path += "" + (path.includes('?') ? '&' : '?') + queryString;
            }
            return path;
        };
    }
    return Endpoint;
}());
exports.Endpoint = Endpoint;
